### ISSUE TITLE
CustomFade will jitter when blending with a mixer state

### ISSUE THREAD

**Issue by bysameng** [KybernetikGames/animancer#147]
# Environment

- Animancer Version Number: 7.1
- Animancer Pro or Lite: Pro
- Unity Version: 2020.3.17 LTS
- Platform: macOS

# Description

When blending between a mixer state and a normal ClipTransition state, and applying a CustomFade, Animancer 7 will jitter uncontrollably.

This worked in Animancer 6, so I believe it's a regression.

# Reproduction

Steps to reproduce the bug:

Attached repro.

1. Open animancer-example scene LinearBlending.unity
2. Use edited version of animancer-example script LinearMixerLocomotion.cs
[LinearMixerLocomotion.cs.zip](https://github.com/KybernetikGames/animancer/files/7073526/LinearMixerLocomotion.cs.zip)

3. In the new Action field, select Humanoid-GolfSwing
• Set "Fade Duration" = 0.4x
• Set "End Time" = 0.8x
• Set "Do Easing Function" = true

<img width="635" alt="Screen Shot 2021-08-29 at 9 15 10 PM" src="https://user-images.githubusercontent.com/5396947/131272439-7af556ab-d4f9-4ab5-996c-a64174c59b73.png">

4. Enter Play Mode
5. Click anywhere on the screen to play GolfSwing animation.
6. Animation will Jitter.
7. Set "Do Easing Function" = false
8. Click to play animation
9. Animation will not Jitter.

**Comment by bysameng** [comment]
https://user-images.githubusercontent.com/5396947/131272582-c75be85f-3def-460d-84a1-ef3294eb7688.mov

**Comment by bysameng** [comment]
Here is the relevant parts of the edited script:

`    public sealed class LinearMixerLocomotion : MonoBehaviour
    {

        [SerializeField] private AnimancerComponent _Animancer;
        [SerializeField] private LinearMixerTransitionAsset.UnShared _Mixer;
        [SerializeField] private ClipTransition _Action;

        public bool doEasingFunction;

        private void Update()
        {
            if (Input.GetMouseButtonDown(0))
            {
                var state = _Animancer.Play(_Action);
                if (state.FadeSpeed > 0)
                {
                    if (doEasingFunction)
                    {
                        CustomFade.Apply(_Animancer, Easing.Function.SineInOut);
                        Debug.Log("Applying fade ");
                    }
                }
                state.Events.OnEnd += () =>
                {
                    var idle = _Animancer.Play(_Mixer);
                    if (doEasingFunction)
                    {
                        CustomFade.Apply(_Animancer, Easing.Function.SineInOut);
                    }
                };
            }
        }
    }`

**Comment by KybernetikGames** [comment]
It seems to only happen when the `AnimancerComponent` is visible in the Inspector (or at least, it's more prominent then). I suspect you might have just not noticed it in the earlier version since I can't think of any change that should affect something like this and the weights applied by the `CustomFade` seem to be correct.

I'm not sure why it only happens when a `CustomFade` is used, but I think the real problem is actually the fact that Root Motion is being used while the Animator's Update Mode is set to Normal. Disabling Root Motion or setting the mode to Animate Physics (before entering Play Mode) seems to avoid the issue.

Also, you shouldn't be adding events like that. When you play a transition it gives the transition's events to the state so if you modify the events after playing it the modification will affect the transition and next time you play it you'll be performing the same modification again (in this case, adding another copy of the end event). So when using transitions you should only be configuring the events once on startup instead of whenever you play them.

**Comment by bysameng** [comment]
Hey thanks for looking into this so quickly!

In my own project, which worked well on Animancer 6, I am not using Root Motion at all. So I don't believe that's the case.

There's even a really weird scale inflation of my character when using any CustomFade. The character's torso blows up. And the legs go in weird directions.

I haven't changed anything else in my project other than updating to 7.1. It worked perfectly on v6 with no jitters.

Thanks for the tip about the end-event! I will avoid doing that.

**Comment by KybernetikGames** [comment]
I think I found the issue. Animancer v7.0 simplified the way things get updated so states are in the same list as the `CustomFade`. In this case, one state is updating first to apply its `Weight` to the Playable Graph, then the `CustomFade` modifies the `Weight` of both states, then the second state applies the modified `Weight` but the first state is still only at the previous value so they don't add up to 1 and the total varies slightly each frame according to the delta time. That gives a result slightly between a proper pose and [the default hunched over pose of a Humanoid rig](https://kybernetik.com.au/animancer/docs/help/debugging/#rig-types), which means the root position will flicker up and down.

To fix it, go to the `CustomFade` script, find the 3 places it calls `SetWeight`, and add a call to `ApplyWeight` right afterwards, i.e.:
```
                _Target.Node.SetWeight(Mathf.LerpUnclamped(_Target.StartingWeight, _Target.Node.TargetWeight, weight));
                _Target.Node.ApplyWeight();

...

                    node.Node.SetWeight(node.StartingWeight * weight);
                    node.Node.ApplyWeight();

...

            node.SetWeight(weight);
            node.ApplyWeight();
```

**Comment by bysameng** [comment]
That fixed it!!! Silky smooth transitions are back with no jitter.

Thank you. I'm repurchasing Animancer again on itch as my thanks. What amazing support ❤️

**Comment by KybernetikGames** [comment]
Yeah, I keep bugs open until the fix is in the latest version.

**Comment by KybernetikGames** [comment]
This issue is fixed in [Animancer v7.2](https://kybernetik.com.au/animancer/docs/changes/animancer-v7-2).
